### PR TITLE
PUBDEV-8496: show original values on ice lines

### PR DESCRIPTION
**Pull request by koniecsveta**
https://h2oai.atlassian.net/browse/PUBDEV-8496

Do you think there should be a parameter to switch this off/on? Currently there is no switcher.

python examples:

binary:
bin_prostate_VOL

regression (no NAs/ NAs/ ICE by categorical variable):
reg_titanic_fare
reg_titanic_body
reg_titanic_cabin_type

multiclass:
mult_iris_Petal Length

R examples:

binary:
RplotprostateID
RplotprostatePSA

regression (numerical/categorical):
Rplottitanicage
Rplottitanicpclass

multiclass:

RplotirisSepal Length

thanks to this, there will be shown original values on ice lines.

### CODE REVIEW

File: h2o-py/h2o/explanation/_explain.py [hunk 1]
```diff
@@ -1326,6 +1330,9 @@ def ice_plot(
             else:
                 plt.plot(tmp[encoded_col], tmp["mean_response"], color=colors[i],
                          label="{}th Percentile".format(i * 10))
+            plt.scatter(frame.as_data_frame()[column][index], model.predict(frame).as_data_frame()['predict'][index],
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1333]: It seems rather wasteful to make a prediction on a whole frame in a for loop, especially if only a single value is used always. 
1) Isn't it sufficient to use the PDP values? To me, as an outside spectator, it seems that `partial_plot` with `row_index` specified would be doing the same thing and the prediction for this particular variable value should be matching. Please correct me if that's a wrong assumption
2) If the value from `partial_plot` is wrong, then would it make sense to make just a single prediction on a whole frame, and just reuse it?

**Review by johnygomez — COMMENTED** [review]
@koniecsveta I find it rather confusing that the observation points are not lying on the ICE line. How is that possible? I apologize if the question is irrelevant, or the answer obvious, and this is expected. Thanks

File: h2o-py/h2o/explanation/_explain.py (hunk 1 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1333]: For a single point (categorical variable) it is matching. For continuous, the point can be outside of its underlying ice line since line is created as an interpolation of a couple of points and that can be imperfect for small nbins. I was discussing this with @tomasfryda and it seemed to us more proper to paint exact predicted value than interpolated value.

prediction should be only 1 and reused, thx 👍

Comment (johnygomez) [reply h2o-py/h2o/explanation/_explain.py:1333]: @koniecsveta that's not entirely clear to me. Are you saying that predictions are computed on multiple rows and ICE for a specific row is interpolated? Or are you referring to some line smoothening interpolation function?

Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1333]: @johnygomez to some line smoothening interpolation function

Comment (johnygomez) [reply h2o-py/h2o/explanation/_explain.py:1333]: I see, would it make sense to try different interpolation functions then, so that they match points more closely? I'm sure you analyzed that for a longer time, it's just that at first glance it looks rather confusing.

Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1333]: when I set nbins to 100, what is the required default value for nbins configurable by the parameter which is to be added as another ICE improvement (https://h2oai.atlassian.net/browse/PUBDEV-8494), I get this much better plot:
age_both

I think I could set nbins=100 already in this PR (currently we are using unconfigurable 20 which is simply insufficient) and add configurability in underlying ticket, WDYT?

Comment (johnygomez) [reply h2o-py/h2o/explanation/_explain.py:1333]: IMHO setting this to a fixed value won't solve the issue, as it strongly depends on the data. It's however better than before and can be addressed later ofc. When looking at the code, however, are you sure there really is any interpolation? If so, where does it happen?

Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1333]: Agree.. I meant ggplot2::geom_line() simple linear interpolation. It is only as descriptive as many points you give it. Currently there is no advanced logic dealing with interpolation in h2o ice_plot code.

Comment (johnygomez) [reply h2o-py/h2o/explanation/_explain.py:1333]: @koniecsveta but that's only for R, I don't see the interpolation done in matplotlib, so I'd assume there is none then. Still, the points are away from the line.

When looking at the R example, however, it also looks fishy as the ICE line seems to be linear, yet the points are far away. How is that possible? That cannot be explained by the interpolation.
image

Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1333]: 👍 looking into it

Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1333]: @johnygomez I've found the issue: df3a79e 

The problem was with indexing. Java indexes from 0 and R from 1, hence if I calculate ICE points on Java backend for let's say `i`, the corresponding row for original value is `i + 1`, not `i` :woman_facepalming:

So currently R output is consistent with py output:
Rplotage

File: h2o-r/h2o-package/R/explain.R [hunk 2]
```diff
@@ -2420,29 +2421,63 @@ h2o.ice_plot <- function(model,
...
+            "Original observation of '",
+            column,
+            "' for ",
+            percentile_str,
+            " is ",
+            interval,
+            ". Ploting of NAs is not yet supported.",
+            sep = ""
+          )
+        print(message)
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/explain.R:2458]: It's probably better to use `warning(message)` (or `cat(message, "\n")` but I would prefer warning since the user can influence its behavior via `options(warn = ... )`).

File: h2o-py/h2o/explanation/_explain.py [hunk 3]
```diff
@@ -1305,35 +1308,59 @@ def ice_plot(
...
+                        frame,
+                        cols=[column],
+                        plot=False,
+                        row_index=index,
+                        targets=target,
+                        user_splits=user_splits,
+                    )[0]
+                )
+                plt.scatter(orig_tmp[encoded_col], orig_tmp["mean_response"],
+                            color=[colors[i]], marker='o', s=150, alpha=0.5)
```
Comment (tomasfryda) [inline h2o-py/h2o/explanation/_explain.py:1348]: It would be good to have the "original observation" entry in the legend as is done in R.

Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1348]: [93d3066](https://github.com/h2oai/h2o-3/pull/6020/commits/93d3066903cc11f3d2
cncnfare
d52cb61d55cd07fb810d12)
dsdscabin_type

File: h2o-py/h2o/explanation/_explain.py [hunk 4]
```diff
@@ -1305,35 +1308,59 @@ def ice_plot(
...
-                    nbins=20 if not is_factor else 1 + frame[column].nlevels()[0]
+                    nbins=100 if not is_factor else 1 + frame[column].nlevels()[0],
+                    include_na=True
                 )[0]
             )
+            user_splits = dict()
+            orig_value = frame.as_data_frame()[column][index]
             encoded_col = tmp.columns[0]
+            if _isnan(orig_value):
+                id = np.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0] if is_factor else np.where(np.isnan(tmp[encoded_col]))[0][0]
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1331]: nit/style: I'd tend to avoid usage of `id` name, as it's built-in python function

nit/style: I find really hard to understand what e.g. 
```pythonnp.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0]```
 is doing. Is `".missing(NA)"` some convention for naming the missing values? If so, maybe it would be worth defining it as a constant. Also what happens if there is no such value found, it seems to fail on `IndexError` or so?

File: h2o-py/h2o/explanation/_explain.py [hunk 5]
```diff
@@ -1305,35 +1308,59 @@ def ice_plot(
...
+                orig_null_value = tmp.from_num_to_factor(encoded_col)[tmp[encoded_col][id]] if is_factor else tmp[encoded_col][id]
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1332]: nit/style: Use complete if/else statements instead of conditional expressions, as the cases here are complicated and hard to parse. I.e. use `x = A if y else B` only for simple cases. This is same for previous line

File: h2o-py/h2o/explanation/_explain.py [hunk 6]
```diff
@@ -1305,35 +1308,59 @@ def ice_plot(
...
+                interval = "[{}, {}]".format(orig_null_value,  tmp["mean_response"][id])
+                warnings.warn("Original observation of \"" + encoded_col + "\" for " + percentile_string + " is " + interval + ". Ploting of NAs is not yet supported." )
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1334]: ultranit/style: This is just observation, but there are multiple string interpolation styles used (`.format` vs concatenating). Please consider unifying it. I'm not sure if the codebase needs to stay Python2 compatible, but elegant modern alternative is to use [Python F-strings](https://realpython.com/python-f-strings/)

File: h2o-py/h2o/explanation/_explain.py [hunk 7]
```diff
@@ -1305,35 +1308,59 @@ def ice_plot(
...
                     targets=target,
-                    nbins=20 if not is_factor else 1 + frame[column].nlevels()[0]
+                    nbins=100 if not is_factor else 1 + frame[column].nlevels()[0],
+                    include_na=True
                 )[0]
             )
+            user_splits = dict()
+            orig_value = frame.as_data_frame()[column][index]
             encoded_col = tmp.columns[0]
+            if _isnan(orig_value):
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1330]: nit/style: This whole if block is pretty hard to parse, please consider separating it into a separate function. I'd also suggest adding some unit tests for it, but I'm not sure if you usually write units for such cases, so consider this optional 🙂

File: h2o-py/h2o/explanation/_explain.py [hunk 8]
```diff
@@ -1362,6 +1396,11 @@ def ice_plot(
             plt.savefig(fname=save_plot_path)
         return decorate_plot_result(figure=fig)
 
+def _isnan(value):
+    if isinstance(value, float):
+        return np.isnan(value)
+    else:
+        return False
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1403]: I consider this to be a little fishy because basically anything that is not a _float_ is automatically a _NaN_. It seems that more robust would be to at least try casting the value to float, and only then perform the check.

File: h2o-py/h2o/explanation/_explain.py [hunk 9]
```diff
@@ -1305,35 +1308,59 @@ def ice_plot(
...
+            user_splits = dict()
+            orig_value = frame.as_data_frame()[column][index]
             encoded_col = tmp.columns[0]
+            if _isnan(orig_value):
+                id = np.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0] if is_factor else np.where(np.isnan(tmp[encoded_col]))[0][0]
+                orig_null_value = tmp.from_num_to_factor(encoded_col)[tmp[encoded_col][id]] if is_factor else tmp[encoded_col][id]
+                interval = "[{}, {}]".format(orig_null_value,  tmp["mean_response"][id])
+                warnings.warn("Original observation of \"" + encoded_col + "\" for " + percentile_string + " is " + interval + ". Ploting of NAs is not yet supported." )
+            else:
+                user_splits[column] = [orig_value]
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1336]: QQ: What's the purpose of this `user_splits` variable? It seems it's always the same dict, just the `orig_value` changes , is that right?

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+sys.path.insert(1,"../../")
+import matplotlib
+matplotlib.use("Agg")  # remove warning from python2 (missing TKinter)
+import h2o
+import matplotlib.pyplot
+from tests import pyunit_utils
+from h2o.estimators import *
+
+
+def test_original_values():
```
Comment (johnygomez) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:14]: This test would be actually a great example for test parametrization, as most of the logic is shared! I'm not sure what test framework you use, but e.g. in pytest, this is achieved using [`parametrize` decorator.](https://docs.pytest.org/en/6.2.x/parametrize.html). Consider if something like that is possible also in this codebase 🙂

**Review by johnygomez — COMMENTED** [review]
I have a few stylistic comments, but please take it as optional as I'm not really familiar with your H2O-3 style preferences 🙂

File: h2o-py/h2o/explanation/_explain.py (hunk 8 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1403]: only `orig_value` is passed into this method (`orig_value = frame.as_data_frame()[column][index]`).

`as_data_frame() `reads the dataset as a python local object and it behaves like: 

- if there is a missing value present in certain numerical column, then the resulting  `frame.as_data_frame()` contains this column with type 'float'. Even if it is of type int in `frame`. Because python NaN is defined as a special floating-point value and cannot be converted to any other type than float.

- if there is or is not a missing value present in certain factor column (binary or other enumeration), then the resulting `frame.as_data_frame()` contains this column with type 'object'. On indexes with non-missing values it contains strings with categories, on indexes with missing values it contains NaN (of type float), hence, if the orig_value (= frame.as_data_frame()[column][index]) is not a float, I can safely say it is not a NaN. Being float is a prerequisite to being NaN here. But trying to cast a string value to float, for example float("Class 1"), will certainly result in a ValueError.

File: h2o-py/h2o/explanation/_explain.py (hunk 4 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1331]: yesyes, ".missing(NA)" is a convention for naming missing values in `tmp` in case of factor column. I can put it into constant.

The thing behind this is, that I am looking for the original value to plot it. It can either be a non-missing value or a missing value and I need to calculate respective response value.

- In case it is a non-missing value, I calculate partial_plot with a given `index `and predefined `user_splits`. into `user_splits` I put only the x-axis value of the original value so that I get response value at this point calculated by exactly the same transformations as points of ice line. Previous call with bins calculated points of the ice line - as many points as many bins, this call calculates only one (the original) point, with x-value defined in user_splits.

under the hood it does this:
```
try {
  preds = _model_id.get().score(test, Key.make().toString(), _job, false);
  if (preds == null || preds.numRows() == 0) {  // this can happen if algo will not predict on rows with NAs
    _meanResponse[_pdp_row_index] = Double.NaN;
    _stddevResponse[_pdp_row_index] = Double.NaN;
    _stdErrorOfTheMeanResponse[_pdp_row_index] = Double.NaN;
  } else {
    CalculateWeightMeanSTD calMeansSTD = (_weight_column_index >= 0)?
            getWeightedStat(fr, preds, _predictorColumn):null;
    _meanResponse[_pdp_row_index] = (_weight_column_index >= 0)?calMeansSTD.getWeightedMean()
            :preds.vec(_predictorColumn).mean();
    _stddevResponse[_pdp_row_index] = (_weight_column_index >= 0)?calMeansSTD.getWeightedSigma()
            :preds.vec(_predictorColumn).sigma();
    _stdErrorOfTheMeanResponse[_pdp_row_index] = _stddevResponse[_pdp_row_index]/Math.sqrt(preds.numRows());
  }
```
in PartialDependence.java

if I only used model.score() value for the original value calculation, for example for the binary model I would need to solve cases what to use as a response from scoredd output which looks like this

`predict        p0        p1
1       1 0.3597337 0.6402663`
 and I would need to solve weights etc to everythiing be in accordance with what is already displayed. By calling partial dependence on custom point defined by 'user_splits' I get a value which is in accordance with other ice plot points and I can be sure it is in the same scale with the rest of the plot and gone through the same transformations.


-  In case it is a missing value, I also need to calculate this value separately, but in h2o it is not possible to pass the nan to  'user_splits'. Hence, I calculate nan value right with the other ice line points by adding  `include_na=True`. Based on type of 'column' it is not guaranteed, which line in tmp will contain the result for nan. So I need to find that index and that's what line 
`idx = np.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0] if is_factor else np.where(np.isnan(tmp[encoded_col]))[0][0]` does. The value will be there, because I specified by include_na=True to be calculated, so IndexError can't happen and if it happens, then something got broken in partial dependence calculation and the calculation is corrupted anyway.
Next line 
`tmp.from_num_to_factor(encoded_col)[tmp[encoded_col][idx]] if is_factor else tmp[encoded_col][idx]`
gives me its x-value in tmp, I could use `orig_null_value = '.missing(NA)' if is_factor else np.nan` instead of it. It served me as a proof of correctness of selected row in tmp, will leave it commented there in case some bug tracing will be needed.

File: h2o-py/h2o/explanation/_explain.py (hunk 9 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1336]: this is explained in 1st point of answer to 1st comment about line `pythonnp.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0]`

File: h2o-py/h2o/explanation/_explain.py (hunk 5 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: [77c3d0a](https://github.com/h2oai/h2o-3/pull/6020/commits/77c3d0acca2ed68f6c134d67015734bc7c5af0f3)

File: h2o-py/h2o/explanation/_explain.py (hunk 4 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: [77c3d0a](https://github.com/h2oai/h2o-3/pull/6020/commits/77c3d0acca2ed68f6c134d67015734bc7c5af0f3)

File: h2o-py/h2o/explanation/_explain.py (hunk 6 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: unified 👍  it has to stay python2 compatible 
[77c3d0a](https://github.com/h2oai/h2o-3/pull/6020/commits/77c3d0acca2ed68f6c134d67015734bc7c5af0f3)

File: h2o-py/h2o/explanation/_explain.py (hunk 4 above)
Comment (johnygomez) [reply h2o-py/h2o/explanation/_explain.py]: Hmm I have to admit that this is really confusing and if I was debugging this code later I probably wouldn't have figured out 😓 I expect you agreed on that beforehand, or maybe it's part of the requirements spec idk, but is it really needed to display original points for missing value then? Maybe I'm missing something, but the point itself is then just an interpolation, which hypothetically can be incorrect. Anyways, I don't want to dig into it too much, so it's fine from my perspective then. Thank you for explanation!

Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: It's good that it's discussed 👍  It's not an interpolation. For certain original value I am calculating pdp/ice in one exact point, the result of `orig_tmp = NumpyFrame( model.partial_plot(... user_splits = [orig_value] ...))` is one exact value to paint a point, not multiple values to be connected into a line. In line `preds = _model_id.get().score(test, Key.make().toString(), _job, false);` from the snippet I've sent above `test` frame consist of  1 row -> `preds `consist of  1 row -> `preds.vec(_predictorColumn).mean() `is mean out of 1 value -> `orig_tmp` consist of 1 row -> 1 point to paint, not multiple points to be connected into a line.

The another part of rquirement is to be able to plot not only NAs, but also user-specified invalid values, so the plot will be supposed to have "numerical part" of x-axis and "NA/invalid part" of x-axis // or there will be two plots, one with valid data, second with NA/invalid data depending on what will be doable. But this part will be done in https://h2oai.atlassian.net/browse/PUBDEV-8493.

Screenshot 2022-02-11 at 18 50 44

File: h2o-py/h2o/explanation/_explain.py [hunk 11]
```diff
@@ -1226,6 +1226,33 @@ def pd_multi_plot(
...
+                        index, column, color, percentile_string):
+    PDP_RESULT_FACTOR_NAN_MARKER = '.missing(NA)'
+    user_splits = dict()
+    if _isnan(orig_value):
+        if is_factor:
+            idx = np.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)[PDP_RESULT_FACTOR_NAN_MARKER])[0][0]
+        else:
+            idx = np.where(np.isnan(tmp[encoded_col]))[0][0]
+        # orig_null_value = tmp.from_num_to_factor(encoded_col)[tmp[encoded_col][idx]] if is_factor else tmp[encoded_col][idx]
+        orig_null_value = PDP_RESULT_FACTOR_NAN_MARKER if is_factor else np.nan
```
[the remark this answers is not in the document]
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1239]: In case that original value is NA, it prints a warning with it's value and informs that plotting NAs is not yet supported (will be in PUBDEV-8493) - that's what this block is about. In order to print the value of NA it's needed to find it in `tmp`, because NA values are calculated together with other ice plot points by using `include_na=True`. So `idx` is index of row that contains NA. If column is factor, I am looking for NA in form `PDP_RESULT_FACTOR_NAN_MARKER`, if it's numerical I am looking for NA in form `np.nan`. When I have `idx` I can extract respective y-axis value`tmp["mean_response"][idx]`.

In case it is not NA, it calculates the y-axis value only for that one point into `orig_tmp` and plots it, thats the else block.

I would leave there the commented line for debugging purposes, but can remove if it is annoying.

File: h2o-py/h2o/explanation/_explain.py [hunk 12]
```diff
@@ -1305,35 +1335,40 @@ def ice_plot(
...
                     row_index=index,
                     targets=target,
-                    nbins=20 if not is_factor else 1 + frame[column].nlevels()[0]
+                    nbins=100 if not is_factor else 1 + frame[column].nlevels()[0],
+                    include_na=True
                 )[0]
             )
             encoded_col = tmp.columns[0]
+            _handle_orig_values(frame.as_data_frame()[column][index], is_factor, tmp, encoded_col, plt, target, model,
+                                frame, index, column, colors[i], percentile_string)
```
[the remark this answers is not in the document]
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: 👍 removed

File: h2o-py/h2o/explanation/_explain.py (hunk 7 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: added test_handle_orig_values() to pyunit_ice_plot.py

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 13]
```diff
@@ -39,7 +135,10 @@ def test_display_mode():
...
-    pyunit_utils.standalone_test(test_display_mode)
-else:
-    test_display_mode()
+
+
+pyunit_utils.run_tests([
+    test_original_values,
+    test_handle_orig_values,
+    test_display_mode
+])
```
Comment (johnygomez) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:144]: nit: missing newline

**Review by johnygomez — APPROVED** [review]
LGTM, thanks for the changes! ✌️

**Comment by koniecsveta** [comment]
(update after todays call) todo: original points should be a part of input points for drawing individual line. in this way it always will be part of the line no matter how many points will be calculated and that will become important when we will select observations not as a deciles but by different options and there will be no color differentiation between them.

File: h2o-py/h2o/explanation/_explain.py [hunk 14]
```diff
@@ -1413,6 +1460,11 @@ def _get_response(mean_response, show_logodds):
         return np.log(mean_response / (1 - mean_response))
     else:
         return mean_response
+def _isnan(value):
```
Comment (tomasfryda) [inline h2o-py/h2o/explanation/_explain.py:1463]: [PEP-8](https://www.python.org/dev/peps/pep-0008/#blank-lines) suggests newlines before the function.
```suggestion


def _isnan(value):
```

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 15]
```diff
@@ -1,15 +1,138 @@
...
+        gbm = H2OGradientBoostingEstimator(seed=1234, model_id="my_awesome_model_py" + str(i))
+        gbm.train(y=ys[i], training_frame=train)
+
+        cols_to_test = _get_cols_to_test(train, ys[i])
+        if names_to_extract[i] is not None:
+            if names_to_extract[i] in cols_to_test: cols_to_test.remove(names_to_extract[i])
+
+        _assert_pyplot_was_produced(cols_to_test, gbm, train, target=targets[i])
+
+def test_handle_orig_values():
```
Comment (tomasfryda) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:33]:
```suggestion


def test_handle_orig_values():
```

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 16]
```diff
@@ -1,15 +1,138 @@
...
+    cols_to_test = []
+    for col, typ in train.types.items():
+        for ctt in cols_to_test:
+            if typ == train.types[ctt] or col == y:
+                break
+        else:
+            cols_to_test.append(col)
+    return cols_to_test
+
+def _assert_pyplot_was_produced(cols_to_test, model, train, target=None):
```
Comment (tomasfryda) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:103]:
```suggestion


def _assert_pyplot_was_produced(cols_to_test, model, train, target=None):
```

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 17]
```diff
@@ -1,15 +1,138 @@
...
+    for col in cols_to_test:
+        if target is None:
+            assert isinstance(model.ice_plot(train, col).figure(), matplotlib.pyplot.Figure)
+        else:
+            assert isinstance(model.ice_plot(train, col, target=target).figure(), matplotlib.pyplot.Figure)
+    matplotlib.pyplot.close("all")
+
+def test_display_mode():
```
Comment (tomasfryda) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:111]:
```suggestion


def test_display_mode():
```

**Review by tomasfryda — APPROVED** [review]
Looks good to me! Thank you @koniecsveta !

File: h2o-py/h2o/explanation/_explain.py [hunk 18]
```diff
@@ -1329,34 +1363,41 @@ def ice_plot(
...
+            percentile_string = "{}th Percentile".format(i * 10)
             tmp = NumpyFrame(
                 model.partial_plot(
                     frame,
                     cols=[column],
                     plot=False,
                     row_index=index,
                     targets=target,
-                    nbins=20 if not is_factor else 1 + frame[column].nlevels()[0]
+                    nbins=100 if not is_factor else 1 + frame[column].nlevels()[0],
```
[the remark this answers is not in the document]
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1377]: this is going to be configurable in PUBDEV-8494, 100 is supposed to be required default value, for testing purposes I can decrease it in PR for this jira

**Review by ledell — APPROVED** [review]
Only suggestion: we could call it `show_pd` if we wanted it ti mean "show partial dependence" instead of "show partial dependence plot". However, I think either can work.
